### PR TITLE
Add Demos for SDF

### PR DESCRIPTION
**Pull request by Amronos**
Adds demos for SDF. Linked to [#ros2_control/1763](https://github.com/ros-controls/ros2_control/pull/1763).
These demos should help in testing the PR.

<br>

A couple of things I want an opinion on:
1. Should using URDF or SDF be an argument of the launch file?
2. What should be the name of the launch file?
3. I have put the SDF files in a new `sdf` directory is that okay or should something else be done?

<br>

To test the diff_drive demo for SDF:
`ros2 launch gz_ros2_control_demos diff_drive_example_sdf.launch.py`
`ros2 run gz_ros2_control_demos example_diff_drive`
The simulation should show a behaviour similar to the diff_drive demo for URDF.

### CODE REVIEW

File: gz_ros2_control_demos/launch/diff_drive_example_sdf.launch.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
+# Copyright 2022 Open Source Robotics Foundation, Inc.
```
Comment (ahcorde) [inline gz_ros2_control_demos/launch/diff_drive_example_sdf.launch.py]:
```suggestion
# Copyright 2024 Open Source Robotics Foundation, Inc.
```

**Comment by Amronos** [comment]
@ahcorde Can I get an initial review?

File: gz_ros2_control_demos/launch/diff_drive_example_sdf.launch.py [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+    robot_description_content = Command(
+        [
+            PathJoinSubstitution([FindExecutable(name='xacro')]),
+            ' ',
+            PathJoinSubstitution(
+                [FindPackageShare('gz_ros2_control_demos'),
+                 'sdf', 'test_diff_drive.xacro.sdf']
+            ),
+        ]
+    )
```
Comment (ahcorde) [inline gz_ros2_control_demos/launch/diff_drive_example_sdf.launch.py]: you don't need to do that, the sdf file doesn't have any xacro tag. You should read the file as it's

**Review by ahcorde — CHANGES_REQUESTED** [review]

File: gz_ros2_control_demos/launch/diff_drive_example_sdf.launch.py (hunk 2 above)
Comment (ahcorde) [reply gz_ros2_control_demos/launch/diff_drive_example_sdf.launch.py]: you should be able to use something like `from launch.substitutions import FileContent`

**Comment by Amronos** [comment]
@ahcorde I decided it would be better for the description format to be an argument of the launch file. I did need to use an OpaqueFunction for this (please tell me if there is a simple way), have a look at ed7f047.
Another option would be not using SDF at all in the launch files and instead just having them be present in the `sdf` directory.

**Comment by bmagyar** [comment]
If the purpose of the demo is to have a urdf and sdf description separately demonstrated but in the same use case, I think the separate argument makes sense.

**Review by ahcorde — CHANGES_REQUESTED** [review]
Do you mind to add some docs to the README.md ?

**Review by ahcorde — CHANGES_REQUESTED** [review]
There are conflicts, do you mind to fix it ?

**Comment by Amronos** [comment]
If everything looks good now, I will also add SDF versions for the other demos.

File: doc/index.rst [hunk 3]
```diff
@@ -122,26 +122,33 @@ The mimic joint must not have command interfaces configured in the ``<ros2_contr
...
 ==========================================
 
-In addition to the *ros2_control* tags, a Gazebo plugin needs to be added to your URDF that
+In addition to the *ros2_control* tags, a Gazebo plugin needs to be added to your URDF or SDF that
 actually parses the *ros2_control* tags and loads the appropriate hardware interfaces and
 controller manager. By default the *gz_ros2_control* plugin is very simple, though it is also
 extensible via an additional plugin architecture to allow power users to create their own custom
 robot hardware interfaces between *ros2_control* and Gazebo.
 
+URDF:
```
Comment (ahcorde) [inline doc/index.rst:131]:
```suggestion
URDF:

```

File: doc/index.rst [hunk 4]
```diff
@@ -122,26 +122,33 @@ The mimic joint must not have command interfaces configured in the ``<ros2_contr
...
     <plugin filename="libgz_ros2_control-system.so" name="gz_ros2_control::GazeboSimROS2ControlPlugin">
       <parameters>$(find gz_ros2_control_demos)/config/cart_controller.yaml</parameters>
     </plugin>
   </gazebo>
+SDF:
+.. code-block:: xml
+
+  <plugin filename="libgz_ros2_control-system.so" name="gz_ros2_control::GazeboSimROS2ControlPlugin">
+    <parameters>$(find gz_ros2_control_demos)/config/cart_controller.yaml</parameters>
+  </plugin>
```
Comment (ahcorde) [inline doc/index.rst]:
```suggestion
  </gazebo>

SDF:

.. code-block:: xml

  <plugin filename="libgz_ros2_control-system.so" name="gz_ros2_control::GazeboSimROS2ControlPlugin">
    <parameters>$(find gz_ros2_control_demos)/config/cart_controller.yaml</parameters>
  </plugin>
```

File: doc/index.rst [hunk 5]
```diff
@@ -157,6 +164,7 @@ or via ROS parameters:
 
 Additionally, one can specify a namespace and remapping rules, which will be forwarded to the controller_manager and loaded controllers. Add the following ``<ros>`` section:
 
+URDF:
 .. code-block:: xml
```
Comment (ahcorde) [inline doc/index.rst]:
```suggestion
URDF:

.. code-block:: xml

```

File: doc/index.rst [hunk 6]
```diff
@@ -168,11 +176,21 @@ Additionally, one can specify a namespace and remapping rules, which will be for
       </ros>
     </plugin>
   </gazebo>
+SDF:
+.. code-block:: xml
```
Comment (ahcorde) [inline doc/index.rst]:
```suggestion
  </gazebo>

SDF:

.. code-block:: xml

```

File: doc/index.rst [hunk 7]
```diff
@@ -188,9 +206,10 @@ The *gz_ros2_control* Gazebo plugin also provides a pluginlib-based interface to
 These plugins must inherit the ``gz_ros2_control::GazeboSimSystemInterface``, which implements a simulated *ros2_control*
 ``hardware_interface::SystemInterface``. SystemInterface provides API-level access to read and command joint properties.
 
-The respective GazeboSimSystemInterface sub-class is specified in a URDF model and is loaded when the
+The respective GazeboSimSystemInterface sub-class is specified in a URDF or SDF model and is loaded when the
 robot model is loaded. For example, the following XML will load the default plugin:
 
+URDF:
 .. code-block:: xml
```
Comment (ahcorde) [inline doc/index.rst]:
```suggestion
URDF:

.. code-block:: xml
```

File: doc/index.rst [hunk 8]
```diff
@@ -204,13 +223,26 @@ robot model is loaded. For example, the following XML will load the default plug
       ...
     </plugin>
   </gazebo>
+SDF:
+.. code-block:: xml
```
Comment (ahcorde) [inline doc/index.rst]:
```suggestion
  </gazebo>

SDF:

.. code-block:: xml
```

File: doc/index.rst [hunk 9]
```diff
@@ -204,13 +223,26 @@ robot model is loaded. For example, the following XML will load the default plug
...
+  </plugin>
 
 Set up controllers
 -----------------------------------------------------------
 
 Use the tag ``<parameters>`` inside ``<plugin>`` to set the YAML file with the controller configuration
 and use the tag ``<controller_manager_prefix_node_name>`` to set the controller manager node name.
 
+URDF:
 .. code-block:: xml
```
Comment (ahcorde) [inline doc/index.rst]:
```suggestion
URDF:

.. code-block:: xml
```

File: doc/index.rst [hunk 10]
```diff
@@ -219,6 +251,13 @@ and use the tag ``<controller_manager_prefix_node_name>`` to set the controller
       <controller_manager_prefix_node_name>controller_manager</controller_manager_prefix_node_name>
     </plugin>
   <gazebo>
+SDF:
+.. code-block:: xml
```
Comment (ahcorde) [inline doc/index.rst]:
```suggestion
    <gazebo>

SDF:

.. code-block:: xml
```

File: doc/index.rst [hunk 11]
```diff
@@ -233,6 +272,10 @@ gz_ros2_control_demos
 ==========================================
 
 There are some examples in the *gz_ros2_control_demos* package.
+To specify whether to use URDF or SDF, you can launch the demo in the following way (the default is URDF):
+.. code-block:: shell
+
+  ros2 launch gz_ros2_control_demos ....launch.py description_format:=sdf
```
Comment (ahcorde) [inline doc/index.rst]:
```suggestion
  ros2 launch gz_ros2_control_demos <launch file> description_format:=sdf
```

**Review by ahcorde — CHANGES_REQUESTED** [review]
While I was compiling the documentation of this PR I noticed that the setup is wrong. This PR should configure the docs file properly.

also If you run the documentation command there are some errors

File: doc/index.rst [hunk 12]
```diff
@@ -243,15 +253,18 @@ Use the tag ``<parameters>`` inside ``<plugin>`` to set the YAML file with the c
...
+
 .. code-block:: xml
 
   <gazebo>
     <plugin name="gz_ros2_control::GazeboSimROS2ControlPlugin" filename="libgz_ros2_control-system">
       <parameters>$(find gz_ros2_control_demos)/config/cart_controller.yaml</parameters>
       <controller_manager_prefix_node_name>controller_manager</controller_manager_prefix_node_name>
     </plugin>
-  <gazebo>
+    <gazebo>
```
Comment (ahcorde) [inline doc/index.rst]: restore this line change

**Review by ahcorde — CHANGES_REQUESTED** [review]

**Review by ahcorde — CHANGES_REQUESTED** [review]
There are some issues:

```
gz_ros2_control_demos/launch/ackermann_drive_example.launch.py:39:21: E128 continuation line under-indented for visual indent
gz_ros2_control_demos/launch/ackermann_drive_example.launch.py:40:21: E128 continuation line under-indented for visual indent
gz_ros2_control_demos/launch/ackermann_drive_example.launch.py:91:9: E222 multiple spaces after operator
gz_ros2_control_demos/launch/diff_drive_example.launch.py:39:21: E128 continuation line under-indented for visual indent
gz_ros2_control_demos/launch/diff_drive_example.launch.py:40:21: E128 continuation line under-indented for visual indent
```

File: doc/index.rst (hunk 12 above)
Comment (Amronos) [reply doc/index.rst]: Fixed in commit [38b6dc1](https://github.com/ros-controls/gz_ros2_control/pull/427/commits/38b6dc1725a6b883b46e03a3e7cff6ba3c69a3e5).

**Review by ahcorde — CHANGES_REQUESTED** [review]
is the robot moving ? I can't move the diff_drive or ackermann examples

**Comment by Amronos** [comment]
> is the robot moving ? I can't move the diff_drive or ackermann examples

Yes, the robot does move for me upon running `ros2 run gz_ros2_control_demos example_diff_drive` for diff_drive or `ros2 run gz_ros2_control_demos example_ackermann_drive` for ackermann.
Make sure you have the [sdformat_urdf](https://index.ros.org/p/sdformat_urdf/github-ros-sdformat_urdf/#rolling) package installed and are using a version of `hardware_interface` that has [this](https://github.com/ros-controls/ros2_control/commit/93a2a68f6ab4110b7e5816bce134437780ba3bf3) commit.

**Review by ahcorde — APPROVED** [review]

**Comment by ahcorde** [comment]
The issue that I mentioned is unrelated to this PR, we can merge this.

**Comment by christophfroehlich** [comment]
@mergifyio backport jazzy

**Pull request merged**